keps/sig-cli/kubectl-staging.md

```diff
@@ -0,0 +1,105 @@
+---
+title: Move Kubectl Code into Staging
+authors:
+  - "@seans3"
+owning-sig: sig-cli
+participating-sigs:
+reviewers:
+  - "@pwittrock"
+  - "@liggitt"
+approvers:
+  - "@soltysh"
+editor: "@seans3"
+creation-date: 2019-04-09
+last-updated: 2019-04-09
+see-also:
+  - https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/kubectl-staging.md
+status: provisional
+---
+
+
+# Move Kubectl Code into Staging
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+* [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Testing](#testing)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+We propose to move the `pkg/kubectl` code base into staging. This effort would
+entail moving `k8s.io/kubernetes/pkg/kubectl` to
+`k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/kubectl`. The `pkg/kubectl`
+code would not have to move all at once; it could be an incremental process. When
+the code is moved, `k8s.io/kubernetes/pkg/kubectl` imports would be changed to
+`k8s.io/kubectl/pkg/kubectl`.
+
+## Motivation
+
+Moving `pkg/kubectl` to staging would provide at least three benefits:
+
+1. Over time, kubectl has grown to inappropriately depend on various
+parts of the Kubernetes code base, creating tightly coupled code which is
+difficult to modify and maintain. For example, internal resource types (those
+that are used as the hub in type conversion) have escaped the API Server and
+have been incorporated into kubectl. Code in the staging directories can not
+have dependencies on Kubernetes core code. Moving `pkg/kubectl` to staging would
+prove that `pkg/kubectl` does not contain Kubernetes core dependencies, and it
+would permanently decouple kubectl from these dependencies.
+
+2. Currently, it is difficult to reuse kubectl code. By moving `pkg/kubectl` to
```

Comment:
The ability to reuse kubectl code outside of kubernetes/kubernetes is a long standing requests.  Moving code to staging incrementally unblocks this use case.